### DISCUSSION TITLE
LuajitError: builtin/box/console.lua:557: attempt to index local 'buf' (a string value)

### DISCUSSION THREAD

**Discussion by olegrok** [tarantool/tarantool#4742] [Q&A] [answered]
```bash
tarantool --version
Tarantool Enterprise For Mac 2.3.0-331-g4d3b2e684
Target: Darwin-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
C_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts
CXX_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts
```
```lua
tarantool> x = {}
---
...

tarantool> x.a = 'hello'
---
...

tarantool> debug.setmetatable(x.a, {__call = function() return 'good bye' end})
---
- true
...

tarantool> x.a
LuajitError: builtin/box/console.lua:557: attempt to index local 'buf' (a string value)
fatal error, exiting the event loop
```

**Comment by cyrillos** [comment]
This looks very similar to #4154

**Comment by olegrok** [comment]
> This looks very similar to #4154

I'm not sure because in #4154 the main problem is manipulation with global system functions/tables that are used in console module.

In the current issue I don't affect any system things. All my actions are in "user-space"

**Comment by cyrillos** [comment]
yup, but the symptoms are quite alike. surely I do _not_ close this bug, still investigating

**Comment by olegrok** [comment]
There is also https://github.com/tarantool/tarantool/issues/4617

It's fact users can break the console with strange manipulations)

**Comment by cyrillos** [comment]
yes, all these bugs are in my todo list

**Comment by Mons** [comment, marked as answer]
Metatables on a string (and also numbers and nils) are shared and they implement interface of `string` module.

```lua
tarantool> debug.getmetatable("").__index == require 'string'
---
- true
...

tarantool> debug.getmetatable("a") == debug.getmetatable("b")
---
- true
...
```
You **CAN'T** set metatable to anything arbitrary and expect, that everything will continue to work

> users can break the console

This manipulation breaks not console but the whole tarantool runtime

**Comment by igormunkin** [comment]
@Mons, thanks a lot for your detailed explanation. In addition to it I can only cite [Lua Reference Manual](https://www.lua.org/manual/5.1/manual.html#2.8):
> Tables and full userdata have individual metatables (although multiple tables and userdata can share their metatables). Values of all other types share one single metatable per type; that is, there is one single metatable for all numbers, one for all strings, etc.

Closing with wontfix due to luajit misuse.

**Comment by olegrok** [comment]
@igormunkin @Mons Thanks!
